base/tuple.jl

```diff
@@ -14,6 +14,11 @@ true
 """
 NTuple
 
+# convenience function for extracting N from a Tuple (if defined)
+# else return `nothing` for anything else given (such as Vararg or other non-sized Union)
+_counttuple(::Type{<:NTuple{N,Any}}) where {N} = N
+_counttuple(::Type) = nothing
```

Comment:
why return `nothing` here instead of `0`? I guess there are cases where we need to distinguish between Tuple length and non-Tuple length?

Reply:
I guess I could return something of like-type instead `-1`. This would let me use `notnothing` or `something` around the uses of it to give a more explicit error (on master, currently causes a stackoverflow)